### PR TITLE
HWKAPM-771 Fix issue where client 'service name' is being incorrectly…

### PR DESCRIPTION
**Pull request by objectiser**
… associated with the invoked service node

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay Could you review on Monday please?

Reviewer notes:
* This bug occurred due to changing from using FragmentCompletionTime events to the NodeDetails - as FragmentCompletionTime distinguished between 'client' fragments that just included a Producer for a URL from the server fragment that had the Consumer.
* When using just the NodeDetails, we need to perform individual queries based on Producer or !Producer, depending upon whether getting information about a 'client' node or service nodes. Due to this not being distinguished, the service name was being associated with the wrong node in the service dependency diagram.

BTW When running integration tests locally for this fix, I got a failure on the OpenTracing polyglot instrumentation framework test. It didn't seem related, but was happening each time, so I tried it against master and it is now also failing there. Is it possibly related to a recent release of the JS provider?

File: server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearch.java [hunk 1]
```diff
@@ -740,28 +771,26 @@ private void buildCommunicationSummaryStatistics(Map<String, CommunicationSummar
...
-                            .<Filter>get("propertiesServiceFilter")
-                            .getAggregations().get("serviceTerm"));
-                    if (serviceName != null) {
-                        css.setServiceName(serviceName);
-                    }
-
-                    stats.put(id, css);
+                    css.setId(uri);
+                    css.setUri(uri);
+                    stats.put(uri, css);
```
Comment (pavolloffay) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearch.java]: Do I understand it correctly that sometimes you put `id` as a key and sometimes `uri`?  (`id = EndpointUtil.encodeEndpoint(uri,
                          operationBucket.getKey());`)

File: server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearch.java [hunk 2]
```diff
@@ -654,6 +654,25 @@ private void buildCommunicationSummaryStatistics(Map<String, CommunicationSummar
...
+    /**
+     * This method adds node information to the communication summary nodes constructed based on the
+     * communication details.
+     *
+     * @param stats The communication summary stats
+     * @param index The index
+     * @param criteria The query criteria
+     * @param addMetrics Whether to add metrics or just discover any missing nodes
+     * @param clients Whether node information should be located for clients (i.e. fragments with
+     *                                  top level Producer node)
```
Comment (pavolloffay) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearch.java]: By fragment do you mean trace fragment? Then the producer can be any node in the trace fragment

Comment (objectiser) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearch.java]: Yes trace fragment. In this particular case, it is looking for a trace fragment that only has a single Producer node - so a special case where the fragment represents a call out to a service.

File: server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearch.java (hunk 1 above)
Comment (objectiser) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearch.java]: The map key (and id field used in the communication summary stats object) is the endpoint (i.e. uri/op). In this particular instance, there is no operation, so the endpoint (and therefore id) is just the URI.

**Comment by objectiser** [comment]
retest this please

**Comment by objectiser** [comment]
@pavolloffay ready for re-review.

**Comment by objectiser** [comment]
@pavolloffay Thanks!

**Pull request merged**